File: src/main/kotlin/schemas/manifest/DefaultLocaleManifest.kt
package schemas.manifest

import data.PreviousManifestData
import data.locale.Tags
import data.shared.Locale
import github.GitHubDetection
import github.ReleaseNotesFormatter
import github.ReleaseNotesFormatter.cutToCharLimitWithLines
import io.ktor.http.Url
import kotlinx.serialization.Contextual
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import network.WebPageScraper
import schemas.AdditionalMetadata
import schemas.SchemaType
import schemas.Schemas

/**
 * A representation of a multiple-file manifest representing a default app metadata in the OWC. v1.5.0
 */
@Suppress("unused")
@Serializable
data class DefaultLocaleManifest(
    val packageIdentifier: String,
    val packageVersion: String,
    val packageLocale: String,
    val publisher: String,
    @Contextual val publisherUrl: Url? = null,
    @Contextual val publisherSupportUrl: Url? = null,
    @Contextual val privacyUrl: Url? = null,
    val author: String? = null,
    val packageName: String,
    @Contextual val packageUrl: Url? = null,
    val license: String,
    @Contextual val licenseUrl: Url? = null,
    val copyright: String? = null,
    @Contextual val copyrightUrl: Url? = null,
    val shortDescription: String,
    val description: String? = null,
    val moniker: String? = null,
    val tags: List<String>? = null,
    val agreements: List<Agreement>? = null,
    val releaseNotes: String? = null,
    @Contextual val releaseNotesUrl: Url? = null,
    val purchaseUrl: String? = null,
    val installationNotes: String? = null,
    val documentations: List<Documentation>? = null,
    val icons: List<Icon>? = null,
    val manifestType: String,
    val manifestVersion: String = Schemas.MANIFEST_VERSION
) : Manifest() {
    @Serializable
    data class Agreement(
        val agreementLabel: String? = null,
        val agreement: String? = null,
        @Contextual val agreementUrl: Url? = null
    )

    @Serializable
    data class Documentation(
        val documentLabel: String? = null,
        @Contextual val documentUrl: Url? = null
    )

    @Serializable
    data class Icon(
        @Contextual val iconUrl: Url,
        val iconFileType: IconFileType?,
        val iconResolution: IconResolution? = null,
        val iconTheme: IconTheme? = null,
        val iconSha256: String? = null
    ) {
        enum class IconFileType {
            PNG,
            JPEG,
            ICO
        }

        enum class IconResolution {
            CUSTOM,
            @SerialName("16x16") SIZE16,
            @SerialName("20x20") SIZE20,
            @SerialName("24x24") SIZE24,
            @SerialName("30x30") SIZE30,
            @SerialName("32x32") SIZE32,
            @SerialName("36x36") SIZE36,
            @SerialName("40x40") SIZE40,
            @SerialName("48x48") SIZE48,
            @SerialName("60x60") SIZE60,
            @SerialName("64x64") SIZE64,
            @SerialName("72x72") SIZE72,
            @SerialName("80x80") SIZE80,
            @SerialName("96x96") SIZE96,
            @SerialName("256x256") SIZE256
        }

        enum class IconTheme {
            DEFAULT,
            LIGHT,
            DARK,
            HIGH_CONTRAST
        }
    }

    override fun toString() = Schemas.buildManifestString(
        manifest = this,
        rawString = EncodeConfig.yamlDefault.encodeToString(serializer = serializer(), value = this)
    )

    companion object {
        /**
         * Returns the name of the YAML file containing the localized manifest for the given package identifier and default
         * locale.
         *
         * @param identifier the package identifier to get the manifest name for
         * @param defaultLocale the default locale to get the manifest name for, if available
         * @param previousDefaultLocale the previously set default locale, if any
         * @return a string representing the name of the YAML file containing the localized manifest for the given
         * identifier and default locale
         */
        fun getFileName(
            identifier: String,
            defaultLocale: String? = null,
            previousDefaultLocale: String?
        ) = "$identifier.locale.${defaultLocale ?: previousDefaultLocale ?: Locale.DEFAULT_LOCALE}.yaml"

        private fun getBase(
            packageIdentifier: String,
            packageVersion: String,
            publisher: String,
            packageName: String,
            license: String,
            shortDescription: String,
            previousDefaultLocaleManifest: DefaultLocaleManifest?
        ): DefaultLocaleManifest = previousDefaultLocaleManifest ?: DefaultLocaleManifest(
            packageIdentifier = packageIdentifier,
            packageVersion = packageVersion,
            packageLocale = Locale.DEFAULT_LOCALE,
            publisher = publisher,
            packageName = packageName,
            license = license,
            shortDescription = shortDescription,
            manifestType = SchemaType.DEFAULT_LOCALE,
            manifestVersion = Schemas.MANIFEST_VERSION
        )

        suspend fun create(
            packageIdentifier: String,
            packageVersion: String,
            defaultLocale: String? = null,
            license: String,
            licenseUrl: Url? = null,
            author: String? = null,
            packageName: String,
            publisher: String,
            publisherUrl: Url? = null,
            packageUrl: Url? = null,
            copyright: String? = null,
            copyrightUrl: Url? = null,
            shortDescription: String,
            moniker: String? = null,
            tags: List<String>? = null,
            releaseNotesUrl: Url? = null,
            manifestOverride: String? = null,
            description: String? = null,
            pageScraper: WebPageScraper?,
            gitHubDetection: GitHubDetection?,
            additionalMetadata: AdditionalMetadata? = null,
            previousManifestData: PreviousManifestData?
        ): DefaultLocaleManifest {
            val parameterLocaleMetadata = additionalMetadata?.locales?.find {
                it.name.equals(other = defaultLocale, ignoreCase = true)
            }
            val previousDefaultLocaleManifest = previousManifestData?.defaultLocaleManifest
            return getBase(packageIdentifier, packageVersion, publisher, packageName, license, shortDescription, previousDefaultLocaleManifest).copy(
                packageIdentifier = packageIdentifier,
                packageVersion = packageVersion,
                packageLocale = defaultLocale
                    ?: previousManifestData?.versionManifest?.defaultLocale
                    ?: Locale.DEFAULT_LOCALE,
                publisher = publisher,
                publisherUrl = (publisherUrl
                    ?: previousDefaultLocaleManifest?.publisherUrl
                    ?: gitHubDetection?.publisherUrl)
                    ?.ifBlank { null },
                publisherSupportUrl = (previousDefaultLocaleManifest?.publisherSupportUrl
                    ?: gitHubDetection?.publisherSupportUrl
                    ?: pageScraper?.supportUrl?.await())
                    ?.ifBlank { null },
                privacyUrl = (previousDefaultLocaleManifest?.privacyUrl
                    ?: gitHubDetection?.privacyUrl
                    ?: pageScraper?.privacyUrl?.await())
                    ?.ifBlank { null },
                author = (author ?: previousDefaultLocaleManifest?.author)?.ifBlank { null },
                packageName = packageName,
                packageUrl = (packageUrl
                    ?: previousDefaultLocaleManifest?.packageUrl
                    ?: gitHubDetection?.packageUrl)
                    ?.ifBlank { null },
                license = license,
                licenseUrl = (licenseUrl
                    ?: previousDefaultLocaleManifest?.licenseUrl
                    ?: gitHubDetection?.licenseUrl)
                    ?.ifBlank { null },
                copyright = (copyright ?: previousDefaultLocaleManifest?.copyright)?.ifBlank { null },
                copyrightUrl = (copyrightUrl ?: previousDefaultLocaleManifest?.copyrightUrl)?.ifBlank { null },
                shortDescription = shortDescription,
                description = (description ?: previousDefaultLocaleManifest?.description)?.formatDescription(),
                moniker = (moniker ?: previousDefaultLocaleManifest?.moniker)?.ifBlank { null },
                tags = (tags ?: previousDefaultLocaleManifest?.tags)?.take(Tags.validationRules.maxItems)?.ifEmpty { null },
                releaseNotesUrl = (releaseNotesUrl
                    ?: gitHubDetection?.releaseNotesUrl
                    ?: parameterLocaleMetadata?.releaseNotesUrl)
                    ?.ifBlank { null },
                releaseNotes = (gitHubDetection?.releaseNotes ?: parameterLocaleMetadata?.releaseNotes)
                    ?.cutToCharLimitWithLines(ReleaseNotesFormatter.MAX_CHARACTERS)
                    ?.trim(),
                documentations = if (previousDefaultLocaleManifest?.documentations == null) {
                    listOfNotNull(
                        pageScraper?.faqUrl?.await()?.let {
                            Documentation(documentLabel = "FAQ", documentUrl = it)
                        }
                    )
                } else {
                    previousDefaultLocaleManifest.documentations
                }.ifEmpty { null },
                manifestType = SchemaType.DEFAULT_LOCALE,
                manifestVersion = manifestOverride ?: Schemas.MANIFEST_VERSION
            )
        }

        private inline fun Url.ifBlank(defaultValue: () -> Url?): Url? = if (this == Url("")) defaultValue() else this

        private fun String?.formatDescription() = this?.replace(Regex("([A-Z][a-z].*?[.!?]) ?(?=\$|[A-Z])"), "$1\n")
            ?.lines()
            ?.joinToString("\n") { it.trim() }
            ?.trim()
            ?.ifBlank { null }
    }
}

File: src/manifests/default_locale_manifest.rs
use const_format::formatc;
use serde::{Deserialize, Serialize};
use serde_with::skip_serializing_none;
use std::collections::BTreeSet;
use std::mem;
use url::Url;

use crate::github::github_client::GitHubValues;
use crate::manifests::ManifestTrait;
use crate::types::author::Author;
use crate::types::copyright::Copyright;
use crate::types::description::Description;
use crate::types::installation_notes::InstallationNotes;
use crate::types::language_tag::LanguageTag;
use crate::types::license::License;
use crate::types::manifest_type::ManifestType;
use crate::types::manifest_version::ManifestVersion;
use crate::types::moniker::Moniker;
use crate::types::package_identifier::PackageIdentifier;
use crate::types::package_name::PackageName;
use crate::types::package_version::PackageVersion;
use crate::types::publisher::Publisher;
use crate::types::release_notes::ReleaseNotes;
use crate::types::sha_256::Sha256String;
use crate::types::short_description::ShortDescription;
use crate::types::tag::Tag;
use crate::types::urls::copyright_url::CopyrightUrl;
use crate::types::urls::license_url::LicenseUrl;
use crate::types::urls::package_url::PackageUrl;
use crate::types::urls::publisher_support_url::PublisherSupportUrl;
use crate::types::urls::publisher_url::PublisherUrl;
use crate::types::urls::release_notes_url::ReleaseNotesUrl;

#[skip_serializing_none]
#[derive(Serialize, Deserialize, Default)]
#[serde(rename_all = "PascalCase")]
pub struct DefaultLocaleManifest {
    pub package_identifier: PackageIdentifier,
    pub package_version: PackageVersion,
    pub package_locale: LanguageTag,
    pub publisher: Publisher,
    pub publisher_url: Option<PublisherUrl>,
    pub publisher_support_url: Option<PublisherSupportUrl>,
    pub privacy_url: Option<Url>,
    pub author: Option<Author>,
    pub package_name: PackageName,
    pub package_url: Option<PackageUrl>,
    pub license: License,
    pub license_url: Option<LicenseUrl>,
    pub copyright: Option<Copyright>,
    pub copyright_url: Option<CopyrightUrl>,
    pub short_description: ShortDescription,
    pub description: Option<Description>,
    pub moniker: Option<Moniker>,
    pub tags: Option<BTreeSet<Tag>>,
    pub agreements: Option<BTreeSet<Agreement>>,
    pub release_notes: Option<ReleaseNotes>,
    pub release_notes_url: Option<ReleaseNotesUrl>,
    pub purchase_url: Option<Url>,
    pub installation_notes: Option<InstallationNotes>,
    pub documentations: Option<BTreeSet<Documentation>>,
    pub icons: Option<BTreeSet<Icon>>,
    pub manifest_type: ManifestType,
    #[serde(default)]
    pub manifest_version: ManifestVersion,
}

impl ManifestTrait for DefaultLocaleManifest {
    const SCHEMA: &'static str = formatc!(
        "https://aka.ms/winget-manifest.defaultLocale.{}.schema.json",
        ManifestVersion::DEFAULT
    );
    const TYPE: ManifestType = ManifestType::DefaultLocale;
}

#[skip_serializing_none]
#[derive(Serialize, Deserialize, Ord, PartialOrd, Eq, PartialEq)]
#[serde(rename_all = "PascalCase")]
pub struct Agreement {
    #[serde(rename = "AgreementLabel")]
    pub label: Option<String>,
    #[serde(rename = "Agreement")]
    pub text: Option<String>,
    #[serde(rename = "AgreementUrl")]
    pub url: Option<Url>,
}

#[skip_serializing_none]
#[derive(Serialize, Deserialize, Ord, PartialOrd, Eq, PartialEq)]
#[serde(rename_all = "PascalCase")]
pub struct Documentation {
    pub document_label: Option<String>,
    pub document_url: Option<Url>,
}

#[skip_serializing_none]
#[derive(Serialize, Deserialize, Ord, PartialOrd, Eq, PartialEq)]
#[serde(rename_all = "PascalCase")]
pub struct Icon {
    #[serde(rename = "IconUrl")]
    pub url: Url,
    #[serde(rename = "IconFileType")]
    pub file_type: Option<IconFileType>,
    #[serde(rename = "IconResolution")]
    pub resolution: Option<IconResolution>,
    #[serde(rename = "IconTheme")]
    pub theme: Option<IconTheme>,
    #[serde(rename = "IconSha256")]
    pub sha_256: Option<Sha256String>,
}

#[derive(Serialize, Deserialize, Ord, PartialOrd, Eq, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum IconFileType {
    Png,
    Jpeg,
    Ico,
}

#[derive(Serialize, Deserialize, Ord, PartialOrd, Eq, PartialEq)]
pub enum IconResolution {
    Custom,
    #[serde(rename = "16x16")]
    Size16,
    #[serde(rename = "20x20")]
    Size20,
    #[serde(rename = "24x24")]
    Size24,
    #[serde(rename = "30x30")]
    Size30,
    #[serde(rename = "32x32")]
    Size32,
    #[serde(rename = "36x36")]
    Size36,
    #[serde(rename = "40x40")]
    Size40,
    #[serde(rename = "48x48")]
    Size48,
    #[serde(rename = "60x60")]
    Size60,
    #[serde(rename = "64x64")]
    Size64,
    #[serde(rename = "72x72")]
    Size72,
    #[serde(rename = "80x80")]
    Size80,
    #[serde(rename = "96x96")]
    Size96,
    #[serde(rename = "256x256")]
    Size256,
}

#[derive(Serialize, Deserialize, Ord, PartialOrd, Eq, PartialEq)]
pub enum IconTheme {
    Default,
    Light,
    Dark,
    HighContrast,
}

impl DefaultLocaleManifest {
    pub fn update(
        &mut self,
        package_version: &PackageVersion,
        github_values: &mut Option<GitHubValues>,
    ) {
        self.package_version.clone_from(package_version);
        if self.publisher_url.is_none() {
            self.publisher_url = github_values
                .as_mut()
                .map(|values| mem::take(&mut values.publisher_url));
        }
        if self.publisher_support_url.is_none() {
            self.publisher_support_url = github_values
                .as_mut()
                .and_then(|values| values.publisher_support_url.take());
        }
        if self.package_url.is_none() {
            self.package_url = github_values
                .as_ref()
                .map(|values| values.package_url.clone());
        }
        if let Some(github_license) = github_values
            .as_mut()
            .and_then(|values| values.license.take())
        {
            self.license = github_license;
        }
        if let Some(github_license_url) = github_values
            .as_mut()
            .and_then(|values| values.license_url.take())
        {
            self.license_url = Some(github_license_url);
        }
        if self.tags.is_none() {
            self.tags = github_values
                .as_mut()
                .and_then(|values| values.topics.take());
        }
        self.release_notes = github_values
            .as_mut()
            .and_then(|values| values.release_notes.take());
        self.release_notes_url = github_values
            .as_mut()
            .and_then(|values| values.release_notes_url.take());
        self.manifest_type = Self::TYPE;
        self.manifest_version = ManifestVersion::default();
    }
}
